Add podman support via new [-p | -d] flags

d1sgr4c3 opened the pull request.

vpetrog committed:
Dockerfile: Add rsync to the container

When working with linux, kernel headers are often required when
building out of tree modules or various userspace components, or the
kernel selftests.

The kernel has a makefile target `headers' for this [1].
However, this makefile target uses `rsync`, which is not currently part
of the container.

Add rsync to the installed binaries of the containers, to support
installing linux headers.

[1] https://www.kernel.org/doc/html/latest/dev-tools/kselftest.html

Signed-off-by: Evangelos Petrongonas <vpetrog@ieee.org>

vpetrog committed:
build_linux: Remove manual shell redirection

If the container is not running in interactive mode, '2>&1' is appended
to the end of the container command. This is problematic. When
`subprocess.Popen()` is called with a list of arguments, each item on
the list is treated as a literal argument, not as shell syntax.
Therefore, when passing the following arguments for example:
```
-- -C /src/tools/testing/selftests TARGETS=kvm
```
it translates to
```
make O=../out/ -j 12 -C /src/tools/testing/selftests TARGETS=kvm 2>&1"
```
which fails as make can not find a target for `2>&1`.

Remove this cmdline extension as a) it is causing build issues b) is
unnecessary as the redirection is already handled with the `Popen`
`stderr=subprocess.STDOUT` parameter.

Signed-off-by: Evangelos Petrongonas <vpetrog@ieee.org>

Willenst commented on README.md:
```diff
@@ -3,7 +3,7 @@
 [![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
 
-This project provides Docker containers for building the Linux kernel (or other software) with many different compilers.
+This project provides Docker or Podman containers for building the Linux kernel (or other software) with many different compilers.
```
думаю "Docker and Podman" лучше по смыслу подойдет

d1sgr4c3 replied:
согласен
ну там ридми в мегадрафтовом варианте щас

Willenst replied:
Тред тогда пока оставлю, что бы не забылось по итогу :)

Willenst commented on manage_images.py:
```diff
@@ -10,6 +10,7 @@
 import argparse
 import pwd
 import grp
+import shutil
```
доп библиотека точно нужна? ниже отписал как можно без нее

d1sgr4c3 replied:
да

Willenst commented on manage_images.py:
```diff
@@ -115,22 +118,39 @@ def find_id(self):
 
+def identify_runtime():
+    """Check for Docker and Podman in the system PATH using `shutil.which`."""
+    engines = [eng for eng in ['docker', 'podman'] if shutil.which(eng)]
```
В коде уже есть os и subprocess, лучше использовать их чем дополнительную библиотеку. Большее число зависимостей потенциально усложнит будущее сопровождение кода.

d1sgr4c3 replied:
сабпроцесс и os это решение про which
в сапроцесс щапускать which и искать тулу
договорились об этом с Александром (что решение с сабпроцесс --- стрем)

Willenst commented on manage_images.py:
```diff
@@ -115,22 +118,39 @@ def find_id(self):
+    if not engines:
+        sys.exit('[!] ERROR: The container runtime is not installed')
+    if len(engines) == 1:
+        print(f'[+] {engines[0].capitalize()} runtime image engine identified')
```
По коду в основном `docker` идет с маленькой буквы, `.capitalize()` или запись с большой буквы нарушает общий вид. Идея в целом хорошая, но стоит одним отдельным комитом сделать глобальный рестайлинг везде, так будет удобнее внедрить/отклонить решение в основной проект.

d1sgr4c3 replied:
не знаю, это как с точками
с большой просто красивее смотрится + в табличке так

d1sgr4c3 replied:
задал вопрос в чате

Willenst commented on manage_images.py:
```diff
@@ -115,22 +118,39 @@ def find_id(self):
+    if len(engines) == 1:
+        print(f'[+] {engines[0].capitalize()} runtime image engine identified')
+        return engines[0]
+    print('[+] Both Docker and Podman are available. Defaulting to Docker.')
```
Defaulting -- to fail to do something, such as pay a debt, that you legally have to do.

Давай лучше сделаем 'Use docker by default'. Или проще: 'Default: docker'.

d1sgr4c3 replied:
https://translate.yandex.ru/?source_lang=en&target_lang=ru&text=defaulting%20to%20docker

яндекс норм понимает

я выписал этот вопрос на "повестку"
сосредоточимся на других деталях

Willenst replied:
Рекомендую вот таким образом контексты поискать:

https://www.google.com/search?q="defaulting"

Технически пойдет, но чаще люди его в контексте с долгами используют

Willenst commented on manage_images.py:
```diff
@@ -153,15 +173,15 @@ def remove_images(needed_compiler, images):
         print(f'\n[!] WARNING: failed to remove {fail_cnt} container image(s), see the log above')
 
-def list_images(images):
+def list_images(images, runtime):
```
В классе `ContainerImage` есть атрибут `runtime`, можно получать его в том числе внутри самой функции через `ContainerImage.runtime` а не передавать из мейна.

d1sgr4c3 replied:
обрати внимание, что лист имаджес не классовая функция

как ты предлагаешь self вне класса крутить?

сам попробуй туда сделать селф рантайм и ты поймешь в чем твоя ошибка

Willenst replied:
Я не предлагал делать self, реализовать можно вот так:

```
def list_images(images):
    """Show the images and their IDs"""
    print('\nCurrent status:')
    print('-' * 44)
    print(f' {"Ubuntu":<6} | {"Clang":<6} | {"GCC":<6} | {f"{ContainerImage.runtime.capitalize()} Image ID"}')
    print('-' * 44)
    for c in images:
        print(f' {c.ubuntu:<6} | {c.clang:<6} | {c.gcc:<6} | {c.id if c.id else "-"}')
    print('-' * 44)
 ```
 
 и не вести по коду лишний параметр

d1sgr4c3 replied:
согласен, так получше будет

Willenst commented on manage_images.py:
```diff
@@ -190,6 +214,15 @@ def main():
         ContainerImage.quiet = True
 
+    if args.podman and args.docker:
+        sys.exit('[!] ERROR: Invalid combination of image runtime engines')
```
Можно унифицировать со шелл скриптами "Multiple image runtime engines specified". Сейчас можно воспринять будто я их не в том порядке поставил :)

d1sgr4c3 replied:
согласен

Willenst commented on manage_images.py:
```diff
@@ -115,22 +118,39 @@ def find_id(self):
-                return ['docker']
-            if 'permission denied' in out.stderr:
+                return [self.runtime]
+            if self.runtime == 'docker' and 'permission denied' in out.stderr:
```
Я провел дополнительное тестирование и несмотря на рутлес режим в подман тоже могут быть ошибочные случаи которые надо обрабатывать отдельно, вот пример с другими пользователями:

```
┌──(doc㉿DESKTOP-N6SHECD)-[/home/kali/kernel-build-containers]
└─$ id
uid=1004(doc) gid=1004(doc) groups=1004(doc),27(sudo)
                                                                                                                                                                                                                                                                                                                              
┌──(doc㉿DESKTOP-N6SHECD)-[/home/kali/kernel-build-containers]
└─$ podman ps      
Failed to obtain podman configuration: mkdir /run/user/0/libpod: permission denied
```

Происходит если с рута перейти на обычного пользака через `su user`.

Кроме того есть еще такая ошибка:
```
┌──(newuser㉿DESKTOP-N6SHECD)-[~]
└─$ podman ps
WARN[0000] The cgroupv2 manager is set to systemd but there is no systemd user session available 
WARN[0000] For using systemd, you may need to log in using a user session 
WARN[0000] Alternatively, you can enable lingering with: `loginctl enable-linger 1003` (possibly as root) 
WARN[0000] Falling back to --cgroup-manager=cgroupfs    
CONTAINER ID  IMAGE       COMMAND     CREATED     STATUS      PORTS       NAMES
WARN[0000] Failed to add pause process to systemd sandbox cgroup: dbus: couldn't determine address of session bus 
```

если перейти на пользака через `su - user`. Первый случай можно попробовать отледить через поимку uid питоном, он вроде как действует на сессию. Второй кейс вроде не вредит работоспособности, думаю его можно проигнорировать.

TL; DR: проработай работу над ошибками на новосозданных пользавателях, пользавателях с малыми правами, на системе с запрещенными нейспейсами

d1sgr4c3 replied:
все так 
именно для этого у нас есть
```
            sys.exit(f'[!] ERROR: Testing "{" ".join(cmd)}" gives unknown error:\n{out.stderr}')
```
как раз для твогео случая:
> Failed to obtain podman configuration: mkdir /run/user/0/libpod: permission denied

это не наша ошибка, а кривой подман
так что все в порядке

> проработай работу над ошибками на новосозданных пользавателях

мне кажется что это нерелевантно

резюмируя:
**я не хочу подчищать говно за редхат. их софт работает криво -- окей, поймаем ошибкой**

Willenst replied:
> мне кажется что это нерелевантно

linux многопользовательская система, если для какого-то пользака вместо вывода "у вас нехватает прав" выведет вот это:
<img width="2543" height="396" alt="image" src="https://github.com/user-attachments/assets/86689a4d-b754-4618-9705-73b497fb321c" />

то это целиком и полностью на нашей совести а не на редхете,падает без корректной отработки ошибки по итогу наш код.

Репликация у этого следующая:
1. создать нового пользака
2. зайти от него и попытаться что-то собрать

d1sgr4c3 replied:
я не могу открыть фотку,у меня 404 на нее
скинь текстом)

d1sgr4c3 replied:
<img width="2344" height="1389" alt="image" src="https://github.com/user-attachments/assets/713ef9a9-61a4-4b65-9743-79660a58bc41" />

d1sgr4c3 replied:
как я уже говорил
та проверка ловит беды с подманом

d1sgr4c3 replied:
<img width="674" height="1661" alt="image" src="https://github.com/user-attachments/assets/221eb951-f036-4d9e-861f-74114d2e32fe" />

а тут прост работает

Willenst replied:
*Отрабатываем в личке*

d1sgr4c3 replied:
записываю на будущую повестку как
> странное поведение подман при смен юзеров которое ломается во время билда (нельзя поймать в питоне)

Willenst commented on manage_images.py:
```diff
@@ -115,22 +118,39 @@ def find_id(self):
-                return ['sudo', 'docker']
+                return ['sudo', self.runtime]
             sys.exit(f'[!] ERROR: Testing "{" ".join(cmd)}" gives unknown error:\n{out.stderr}')
         except FileNotFoundError:
```
раз мы выявляем наличие движка в другом месте ранее, эта проверка не имеет смысла. Кроме того, можно снять весь трай-эксепшен, по той же причине.

d1sgr4c3 replied:
> раз мы выявляем наличие движка в другом месте ранее,

это в каком?

например, если установлено два движка, то дефолтим в докер

тем не менее вседа есть флажки в скрипте которые оверрайдят автовыбор и делают его пользовательским
именно в этом члучае этот код и работает
посмотри по кавереджу потом
оно нужно если флажки суются грубо говоря

не соглсен -- приведи пример как это можно склеить и напиши набросок кода

Willenst replied:
*Отрабатываем в личке*

d1sgr4c3 replied:
придумали классовы метод, сделаю черновик на будущее сейчас

Willenst commented on start_container.sh:
```diff
@@ -55,6 +47,24 @@ while [[ $# -gt 0 ]]; do
 		set -x
 		shift
 		;;
+	-p|--podman)
```
`-p|--podman)` -> `-p| --podman)`, также и с докером ниже

d1sgr4c3 replied:
согласен

Willenst commented on start_container.sh:
```diff
@@ -71,14 +81,40 @@ while [[ $# -gt 0 ]]; do
+
+output=$($RUNTIME ps 2>&1)
+
+if [ "$RUNTIME" = "podman" ]; then
+	if [ "$EUID" -eq 0 ]; then
+		echo "Dont use podman with sudo, it's rootless!" >&2
+		exit 1
```
здесь выход в случае `sudo podman`/root пользака явный, но в `manage_images.py` можно работать через sudo и не получать даже предупреждений. Стоит унифицировать логику.

d1sgr4c3 replied:
все так
при этом в старт контейнерсе явно печатается пользак и юид
таким образом если кто то пишет баг типа вот я сделал sudo manage images а потом чота не работает то мы ему в его логе укажем что старт контейнерс сигналит айдишку и пользака

если тебе кажется это не достаточным можно сделать дополнительный вывод айдишки при сборке контейнера с подман
как тебе идея? есть другие?

Willenst replied:
Может ту же логику в manage_images.py продублировать и просто также по UID 0 разворачивать людей?

d1sgr4c3 replied:
хорошая идея 
но ты учти что есть люди у которых рут единственный пользак
добавил на повестку

Willenst replied:
Такие люди не смогут использовать в том числе скрипты старта и финиша, поскольку там идет exit

Willenst commented on start_container.sh:
```diff
@@ -1,28 +1,17 @@
-groups | grep '\<docker\>' > /dev/null
-NEED_SUDO=$?
-
-set -eu
```
`set -eu` надо вернуть назад :)

d1sgr4c3 replied:
вернул в обоих скриптах

Willenst replied:
В `start_containers` возможно его на пораньше двинуть? он ошибки ловить должен, в конце уже мало пользы

d1sgr4c3 replied:
присмотрись
он стоит сразу после парсинга аргументов

Willenst commented on finish_container.sh:
```diff
@@ -1,27 +1,42 @@
+	echo "usage: $0 kill/nokill out_dir podman/docker"
+	echo "  kill/nokill -- how to finish: kill the container and then clean up / only clean up"
+	echo "  out_dir -- build output directory used by this container (with container.id file)"
+	echo "  docker/podman -- docker or podman image runtime"
```
Стоит унифицировать с другими выводами по коду, оставив:
```
echo "You have to choose: docker or podman"
echo "usage: $0 kill/nokill out_dir podman/docker"
```

d1sgr4c3 replied:
так этож хелп

зачем его унифицировать с принтами которые об ошибках сигнялят

d1sgr4c3 replied:
echo "You have to choose: docker or podman"

эту штуку мне точно не хочется пиать)
> have to choose: docker or podman

звучит как будто ыбы мне дуло пистолета приставили к голове и заставили выбрать ахах

Willenst replied:
Отсуствие выбранного движка и будет ошибкой, как и с kill/nokill.

Я еще идею тредом ниже закидывал, можно под все ошибки сделать вывод хелп функции, так и унификация будет. Но это уже рефакторинг чужого и лучше отдельным коммитом

d1sgr4c3 replied:
> Отсуствие выбранного движка и будет ошибкой, как и с kill/nokill.

так и должно быть. в скрикптах `strict` все

Willenst replied:
Попробую пример привести: В случае однотипной ошибки неверного параметра выводится разнотипный help message.

Вот пример:
```
$ ./finish_container.sh kill /tmp [wrong]
unexpected state or image runtime engine
usage: ./finish_container.sh kill/nokill out_dir podman/docker
  kill/nokill -- how to finish: kill the container and then clean up / only clean up
  out_dir -- build output directory used by this container (with container.id file)
  docker/podman -- docker or podman image runtime
```

```
$ ./finish_container.sh [kill] /tmp docker
Hey, we gonna use sudo for running docker
You have to choose: kill or nokill
usage: ./finish_container.sh kill/nokill out_dir
```

```
$ ./finish_container.sh                   
usage: ./finish_container.sh kill/nokill out_dir podman/docker
  kill/nokill -- how to finish: kill the container and then clean up / only clean up
  out_dir -- build output directory used by this container (with container.id file)
  docker/podman -- docker or podman image runtime
```

В случае докера help и error месседжи воедино смешались, нужен единый формат: либо сделать kill/nokill как docker/podman, либо наоброт

Willenst commented on finish_container.sh:
```diff
@@ -1,27 +1,42 @@
+
 if [ $ACTION != "kill" -a $ACTION != "nokill" ]; then
 	echo "You have to choose: kill or nokill"
 	echo "usage: $0 kill/nokill out_dir"
```
Необходимо обновить ттут текст на: echo "usage: $0 kill/nokill out_dir podman/docker"

В качестве доп идеи: можно вовсе сделать хелп месседж отдельной функцией и звать его везде по коду

d1sgr4c3 replied:
да, спасибо
добавил
прошелся по этому файлу

Willenst commented on build_linux.py:
```diff
@@ -182,6 +201,15 @@ def main():
 
     print(f'[+] Going to build with: {args.compiler}')
 
+    if args.docker:
+        runtime = 'docker'
+    if args.podman:
+        runtime = 'podman'
+    if args.podman and args.docker:
+        sys.exit('[!] ERROR: Invalid combination of image runtime engines')
+    if not (args.podman or args.docker):
+        runtime = identify_runtime()
+
```
Лучше унифицировать с `manage_images.py`

d1sgr4c3 replied:
согласен

Willenst commented on start_container.sh:
```diff
@@ -1,28 +1,17 @@
 	echo "  -e    add environment variable in the container (may be used multiple times)"
 	echo "  -h    print this help"
 	echo "  -v    enable debug output"
+	echo "  -p    use podman runtime"
+	echo "  -d    use docker runtime"
```
компилятор передается без префикса `-/--`, возможно стоит оставить эту практику и для рантайма контейнеров

d1sgr4c3 replied:
распиши мысль поподробнее пожалуйста

Willenst replied:
У нас есть вот такой usage 
`echo "usage: $0 compiler src_dir out_dir [-n] [-e VAR] [-h] [-v] [-p | -d] [-- cmd with args]"`

т.е. при вызове мы просто пеередаем gcc/clang без преписки `-/--`, я думаю это можно сделать и для рантайма, тогда чуть покрасивее будет его оформить в `build_linux.py`

Willenst commented on build_linux.py:
```diff
@@ -92,6 +92,10 @@ def build_kernel(arch, kconfig, src, out, compiler, make_args):
 
     start_container_cmd = ['bash', os.path.dirname(os.path.abspath(__file__)) + '/start_container.sh',
                                    compiler, src, out_subdir]
+    if runtime == 'podman':
+        start_container_cmd.append('-p')
+    if runtime == 'docker':
+        start_container_cmd.append('-d')
```
start_container.sh умеет работать с полным флагом `--podman` и `--docker`, блок ифов можно убрать и напрямую передавать в `start_container_cmd` рантайм через f'--{runtime}'

d1sgr4c3 replied:
я понял о чем ты, но как ты хочешь рантайм положить если ты его по аргументам дентифиуирешь?
мне кажется это нечитаемо будет

приведи пример пожалуйста

Willenst replied:
Команду выше вот так сделать:
```
    start_container_cmd = ['bash', os.path.dirname(os.path.abspath(__file__)) + '/start_container.sh',
                                   compiler, src, out_subdir, f'--{runtime}']
```

На момент вызова `build_kernel` рантайм уже определен и хранит в себе 'docker'/'podman', с чем как раз может оперировать start_container.sh

d1sgr4c3 replied:
https://github.com/d1sgr4c3/kernel-build-containers/pull/2/commits/69efd3840ea215ec9be7b4430e2eb0d4c3182317

Willenst commented on test_image_mgmt.sh:
```diff
@@ -1,124 +1,172 @@
+	if [ "$RUNTIME" = "podman" ]; then
+		if [ "$EUID" -eq 0 ]; then
+			echo "Dont use podman with sudo, it's rootless!" >&2
+			SUDO_CMD=''
```
Сразу после `SUDO_CMD=''` идет exit.

Кроме того, тут серьезная ошибка в логике. Поскольку тест докера идет первым, переменная SUDO_CMD встает в "sudo" и далее не меняется. Т.к. тесты запускаются от обычного пользавателя, проверка на `"$EUID" -eq 0` не выполняется и для подмана сохраняется SUDO_CMD='sudo', вследствии чего идет падение, когда в подмане идет вызов:
```
sudo podman run -d --rm --name test-running-1 kernel-build-container:gcc-12 tail -f /dev/null
```

d1sgr4c3 replied:
все так, я тоже нашел этот баг у себя
он исправлен в позднем коммите, который я запушил

тоже видел это при тестировании

решил вот так
<img width="3410" height="545" alt="image" src="https://github.com/user-attachments/assets/3868fd9f-8474-47ea-9876-2dd5bcec4320" />

Willenst commented on test_image_mgmt.sh:
```diff
@@ -1,124 +1,172 @@
+SUDO_CMD=''
+
+check_sudo(){
+	set +eux
```
зачем тут этот вызов?

d1sgr4c3 replied:
это просто баш командочка, не вызов
но не буду дотошным
в конце +еух в конце - еух
\+ еух снимает на время выполнения чек судо флажок -е
этот влажок крашит скрипт если есть ошибка
а мы именно в стдэрроре смотрим хватает прав на докер или нет
короче чек судо находится в зеленой бесполетной зоне, освобожден от ошибок и прочей отладочной баш лабуды которую мы используем чтобы брякнуть тест если команда в стэрор какашки насыпет

Willenst replied:
А, спасибо, не знал что так можно, закрываю тред :)

Willenst commented on test_image_mgmt.sh:
```diff
@@ -1,124 +1,172 @@
+	export PATH_BACK="$PATH"
+	export PATH=''
+	/usr/bin/python3 -m coverage run -a --branch manage_images.py -l
+	export PATH="$PATH_BACK"
```
Что делает этот тест?

d1sgr4c3 replied:
тестирует 146,147 строки

Willenst commented on test_image_mgmt.sh:
```diff
@@ -1,124 +1,172 @@
+	export PATH=''
+	/usr/bin/python3 -m coverage run -a --branch manage_images.py -l
+	export PATH="$PATH_BACK"
+	sudo mv /usr/bin/docker /usr/bin/docker_back
+	python3 -m coverage run -a --branch manage_images.py -l
+	python3 -m coverage run -a --branch manage_images.py -l -d
+	sudo mv /usr/bin/docker_back /usr/bin/docker
```
Может лучше менее инвазивно через PATH/alias? Я бы не хотел что бы тест переименовывал мне бинари :)

d1sgr4c3 replied:
у меня не получилось сделать сдесь через PATH
если ты видишь решение -- предложи его пожалуйста, предварительно протетсировав с `coverage`

Willenst commented on test_image_mgmt.sh:
```diff
@@ -1,124 +1,172 @@
+RUNTIME='podman'
+RUNTIME_FLAG='-p'
+runtime_tests
+multiple_runtime
```
вызов `clear_state` перед завершением будет не лишним

d1sgr4c3 replied:
согласен

Willenst commented on build_linux.py:
```diff
@@ -137,13 +141,24 @@ def build_kernel(arch, kconfig, src, out, compiler, make_args):
         sys.exit('[!] Exit by interrupt')
 
+def identify_runtime():
+    """Check for Docker and Podman in the system PATH using `shutil.which`."""
```
Точка в конце излишняя

d1sgr4c3 replied:
почему?
вот она в другом месте
<img width="2055" height="226" alt="image" src="https://github.com/user-attachments/assets/4fcc23f8-4278-48f9-8875-82bce00ff474" />

давай тогда уже определимся -- сносить их или нет

Willenst replied:
Я всецело за полный снос :)

d1sgr4c3 replied:
аналогично

как вкину комент с рефактором -- посигналю

Willenst commented on finish_container.sh:
```diff
@@ -1,27 +1,42 @@
+	echo "Hey, you are running podman as $USERNAME ($(id -u))"
+elif [ "$RUNTIME" = "docker" ]; then
+	if echo "$output" | grep -qi "CONTAINER ID"; then
+		echo "Enough permissions to run docker, sudo is not needed"
```
False positive имеется: если я введу 
`sudo ./finish_container.sh kill /tmp docker` мне сообщат `Enough permissions to run docker, sudo is not needed`, будто я должен перестать использовать sudo.

d1sgr4c3 replied:
ну все правильно написано
дейстивтельно звать судо уже не нужно так как ты от рута выполняешь команду


можешь предложить замену формулировки если считаешь это критичным

Willenst replied:
Может сократим до "Enough permissions to run docker"?

d1sgr4c3 replied:
могу взять `sudo` в кавычки. так мы сделаме акцент на том, что это именно командочка на бекенде

если не согласен -- давай зафиксируем вместе это и придем на синк во вторник, поговорим втроем. ты пишшешь хорошие вещи, но это корнер кейс и превращать прогу в сборище корнер кейсов не хочется. плюс эту формулировку разработали с александром. оставь этот тред непорезолвленным, и мы по ним пройдемся во вотрник

ок?

Willenst commented on manage_images.py:
```diff
@@ -35,6 +37,7 @@ class ContainerImage:
         id (str): container image ID
     """
 
+    runtime = ''
     runtime_cmd = None
```
Рассмотри пожалуйста возможность посоединять логику и использовать только один атрибут, по сути сейчас единственное отличие, что `runtime_cmd` может хранить в себе `sudo`

d1sgr4c3 replied:
я рассматривал, мне кажется так код гораздо менее читаемый станет так как нужно отдельно снаружи класса крутить рантайм

Willenst commented on finish_container.sh:
```diff
@@ -39,15 +54,15 @@ ID=$(awk 'NR==1 {print $1}' $CID_FILE)
-	echo "Killing the docker container $ID"
-	$SUDO_CMD docker kill "$ID"
+	echo "Killing the $RUNTIME container $ID"
+	$SUDO_CMD $RUNTIME kill "$ID"
```
SUDO_CMD задается только в блоке 

```
	elif echo "$output" | grep -qi "permission denied"; then
		echo "Hey, we gonna use sudo for running docker"
		SUDO_CMD="sudo"
	fi
```
Ее надо где-то проинициализировать пустой строкой

d1sgr4c3 replied:
согласен, в другом скрипте задал
тут забыл

Willenst commented on manage_images.py:
```diff
@@ -115,22 +118,39 @@ def find_id(self):
                 sys.exit(f'[!] ERROR: Invalid image "{self.clang_tag}" ' \
                           'without the corresponding GCC tag, please remove it manually')
-        return clang_id
+        if self.runtime == 'docker':
+            return clang_id
+        if self.runtime == 'podman':
+            return clang_id.split()[0] if clang_id else ''
+            # this formatting is necessary due to a podman output
+            # and designed to keep the code working even after a fix
+            # (see https://github.com/containers/podman/issues/25725)
```
1.  нет нужды делать разделение на докер и подман, код для подмана будет корректен и в случае одного id и в случае нескольких
```
>>> a = '123123'
>>> a.split()[0]
'123123'
```

2. Ура, кажется фикс внедрили! Стоит какую то пометку сделать, когда фикс был внедрен, во время следующей работы можно будет убрать эту надстройку

d1sgr4c3 replied:
согласен, склеил

Willenst reviewed (commented):
Привет, отличная работа, отписал по новводениям в коде, помимо них пожалуйста расширь README добавив пункты про создание podman контейнеров

Willenst commented on finish_container.sh:
```diff
@@ -30,7 +32,7 @@ elif [ "$RUNTIME" = "docker" ]; then
 else
 	echo "unexpected state or image runtime engine" >&2
-	echo "usage: $0 kill/nokill out_dir podman/docker"
+	echo "usage: $0 kill/nokill out_dir podman/docker podman/docker"
```
опечатка: 2 раза рантайм просят

d1sgr4c3 committed:
fix: bad links for `llvm` and `lld` if `package-$CLANG_VERSION` installed

this commit should fix strange bug with these packages.while installing clang-<version>,
`apt` also installs `llvm` and `lld` packages. however, `apt` installs `lld-<version>`
and `llvm-<version>` and spawn `/usr/bin/ld.lld-<version>` which is not recognised.

Co-authored-by: Willenst <willenst@gmail.com>

d1sgr4c3 committed:
feature: added podman runtime support, auto-detect podman/docker

according to #31

in this work i implemented a podman image runtime support for kernel-build-containers.
now we can choose image runtime via `-p` or `-d` flags. moreover, this commit introduces
a way to choose runtime image engine without specifying flag `-p` or `-d` directly. bash
scripts written with strict flag logic.

d1sgr4c3 committed:
improvement: implement multiple runtime support for running images

a13xp0p0v committed:
Merge remote-tracking branch 'vpetrog/rsync'

Refers to #36.

a13xp0p0v committed:
manage_images.py: Improve the output

a13xp0p0v committed:
manage_images.py: Simplify the new code

a13xp0p0v committed:
Fix the Docker warnings InvalidDefaultArgInFrom and FromAsCasing for line 2

Here `UBUNTU_VERSION=default` provides an invalid default value to prevent
running `docker build` without setting the `UBUNTU_VERSION` argument.

a13xp0p0v committed:
Fix a mistake in the `update-alternatives` command for `ld.lld`

a13xp0p0v committed:
Merge branch 'llvm-lld'

Refers to #37 and #38.

a13xp0p0v committed:
start_container.sh: Improve the output

a13xp0p0v committed:
start_container.sh: Improve the argument order

a13xp0p0v committed:
Choose Docker/Podman similarly in start_container.sh and manage_images.py

a13xp0p0v committed:
start_container.sh: Introduce RUNTIME_ARGS

a13xp0p0v committed:
start_container.sh: Restore `set -eu`

a13xp0p0v committed:
start_container.sh: Simplify the code that checks whether we need sudo

a13xp0p0v committed:
Report username and UID in Podman mode

That's important because different users have separate Podman containers.

Improve `getpwuid()` and `getgid()` usage by the way.

Improve the output in `start_container.sh` as well.

d1sgr4c3 committed:
refactor(build_linux.py): simplify the runtime choosing

d1sgr4c3 committed:
refactor(build_linux.py): improve the output

d1sgr4c3 committed:
refactor(finish_container.sh): simplify logic, refine arguments order

d1sgr4c3 committed:
refactor(build_linux.py): refine interaction with `finish_container.sh` according to new arguments order

d1sgr4c3 committed:
refactor(finish_container.sh): improve the outputs
